### PR TITLE
[opt](nereids)project sub expression in other condition for nested loop join

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ProjectOtherJoinConditionForNestedLoopJoin.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    @Override
+    public Rule build() {
+        return logicalJoin().when(
+                join -> join.getHashJoinConjuncts().isEmpty()
+                        && !join.isMarkJoin()
+                        && !join.getOtherJoinConjuncts().isEmpty()
+                        && !join.getJoinReorderContext().isExtractSubExpressionForChild()
+        ).then(
+                join -> {
+                    join.getJoinReorderContext().setExtractSubExpressionForChild(true);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ProjectOtherJoinConditionForNestedLoopJoin.java:58]: why need this? just like push hash join conjuncts, when not changed, we return null then the rule will not fall into dead loop

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ProjectOtherJoinConditionForNestedLoopJoin.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                        }
+                        Plan right = join.right();
+                        if (!ctx.rightAlias.isEmpty()) {
+                            List<NamedExpression> newProjects = Lists.newArrayList(right.getOutput());
+                            newProjects.addAll(ctx.rightAlias);
+                            right = new LogicalProject<>(newProjects, right);
+                        }
+                        join = join.withJoinConjuncts(join.getHashJoinConjuncts(),
+                                newOtherConjuncts, join.getJoinReorderContext())
+                                .withChildren(Lists.newArrayList(left, right));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ProjectOtherJoinConditionForNestedLoopJoin.java:84]:
```suggestion
                                .withChildren(ImmutableList.of(left, right));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ProjectOtherJoinConditionForNestedLoopJoin.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                    return join;
+                }
+        ).toRule(RuleType.PROJECT_OTHER_JOIN_CONDITION);
+    }
+
+    private static class ReplacerContext {
+        HashMap<Expression, Alias> aliasMap = new HashMap<>();
+        Set<Slot> leftSlots;
+        Set<Slot> rightSlots;
+        Set<Alias> leftAlias = new HashSet<>();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ProjectOtherJoinConditionForNestedLoopJoin.java:95]: LinkedHashSet?

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ProjectOtherJoinConditionForNestedLoopJoin.java:95]: we do not case about the order here.

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run feut

**Review by starocean999 — APPROVED** [review]

**Review by LiBinfeng-01 — APPROVED** [review]

**Pull request merged**